AzureIntegrationTests share one billing profile

okotsopoulos opened the pull request:
I first converted AzureIntegrationTest to JUnit5.
This was necessary in order to create @BeforeAll and @AfterAll methods to initialize a single Azure billing profile for use in all tests.

Expected benefits:

- If one test in the suite fails to delete its billing profile on tear-down, all other tests in the suite currently fail fast when trying to create datasets with errors like "Cannot reuse existing application deployment integrationapp from profile … with a different profile …" ([example failure mode](https://scans.gradle.com/s/hc4mgqdbdmtc2/tests/overview?outcome=FAILED)).  This change should allow for more precise failure signaling, and increase the efficacy of retries.
- Serve as a model for sharing test initialization within a class, not currently used elsewhere in TDR.  Overall integration test runtime can take 2+ hours, and sharing set-up may help drive that down while incentivizing writing more granular tests when appropriate rather than overloading existing ones.

okotsopoulos committed:
AzureIntegrationTests share one billing profile

I first converted AzureIntegrationTest to JUnit5.
This was necessary in order to create @BeforeAll and @AfterAll methods to initialize a single Azure billing profile for use in all tests.

Expected benefits:

- If one test in the suite fails to delete its billing profile on tear-down, all other tests in the suite currently fail fast when trying to create datasets with errors like "Cannot reuse existing application deployment integrationapp from profile … with a different profile …".  This change should allow for more precise failure signaling, and increase the efficacy of retries.
- Serve as a model for sharing test initialization within a class, not currently used elsewhere in TDR.  Overall integration test runtime can take 2+ hours, and sharing set-up may help drive that down while incentivizing writing more granular tests when appropriate rather than overloading existing ones.

okotsopoulos commented on src/test/java/bio/terra/integration/AzureIntegrationTest.java:
```diff
@@ -117,28 +117,32 @@
-@Category(Integration.class)
-public class AzureIntegrationTest extends UsersBase {
+@Tag(Integration.TAG)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
This annotation is necessary to use `@BeforeAll` and `@AfterAll` annotations on non-static methods.

okotsopoulos committed:
JUnit 5 supports Extensions, not Rules

AzureIntegrationTest had a TestJobWatcher included as a @Rule, which is no longer supported in JUnit 5.
Instead, we can extend the test class with new TestJobWatcherExtension, which is compatible with JUnit 5.
Also refactored two assertions which still used JUnit 4 methods.

okotsopoulos committed:
Revert before merge: only run AzureIntegrationTest suite in CI

To make debugging easier.

okotsopoulos commented on .github/workflows/int-and-connected-test-run.yml:
```diff
@@ -204,16 +204,18 @@ jobs:
           actions_subcommand: 'gradleinttest'
           pgport: ${{ job.services.postgres.ports[5432] }}
           test_to_run: 'testIntegration'
+          test_filter: 'AzureIntegrationTest' ## DO NOT MERGE.
```
I won't check this in, but PSA that this is a supported filter by way of this PR: https://github.com/broadinstitute/datarepo-actions/pull/74

okotsopoulos committed:
Also skip unit tests for now (will not merge)

okotsopoulos committed:
Move blob IO test util initialization back to @BeforeEach methods

To help triangulate source of flaky integration tests.
The valuable part of this work is sharing a billing profile among tests, so removing other possible sources of flakiness.

pshapiro4broad commented on src/test/java/bio/terra/integration/TestJobWatcherUtils.java:
```diff
@@ -0,0 +1,53 @@
+          resource.labels.location="us-central1"
+          resource.labels.cluster_name="integration-master"
+          resource.labels.namespace_name="integration-<intNumber>"
+          labels.k8s-pod/component="integration-<intNumber>-jade-datarepo-api""";
```
Is this missing an extra `"` at the end?
```suggestion
          labels.k8s-pod/component="integration-<intNumber>-jade-datarepo-api"
          """;
```

okotsopoulos replied:
Good eye, thanks!

pshapiro4broad commented on src/test/java/bio/terra/integration/TestJobWatcherUtils.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class TestJobWatcherUtils {
+
+  private static final Logger logger = LoggerFactory.getLogger(TestJobWatcherExtension.class);
```
```suggestion
  private static final Logger logger = LoggerFactory.getLogger(TestJobWatcherUtils.class);
```

pshapiro4broad commented on src/test/java/bio/terra/integration/TestJobWatcherUtils.java:
```diff
@@ -0,0 +1,53 @@
+import org.stringtemplate.v4.ST;
+
+/**
+ * Utility methods for reuse between Rules (JUnit 4) and Extensions (JUnit 5) which aide in
```
```suggestion
 * Utility methods for reuse between Rules (JUnit 4) and Extensions (JUnit 5) which aid in
```

pshapiro4broad commented on src/test/java/bio/terra/integration/AzureIntegrationTest.java:
```diff
@@ -117,28 +116,31 @@
-public class AzureIntegrationTest extends UsersBase {
+@Tag(Integration.TAG)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class AzureIntegrationTest extends UsersBase {
```
Probably not worth fixing here, but I think using inheritance for this this kind of code sharing (especially in a test class) is a poor design. Making `UsersBase` an injectable component with singleton scope would be a way to rewrite this and keep the same functionality. This could be done while keeping all current uses the same.

okotsopoulos committed:
@pshapiro4broad PR feedback

- Lost a closing quote when converting to multiline string
- Typo in docstring

okotsopoulos committed:
Make UsersBase an injectable component

Autowire it into AzureIntegrationTest rather than having the test extend it.
Fix an instance where we mistakenly called UsersBase.steward() in a test where we should have used the steward user initialized in test setup.

okotsopoulos committed:
Fix typos in shell commands documented in render-configs.sh

okotsopoulos committed:
On Azure profile create, register association with deployed app

Previously, the deployed app <-> billing profile association is registered the first time a dataset is created off of the profile.
I believe this is producing a race condition / flaky integration tests using the same billing profile.

- Added new CreateProfileRegisterDeployedApplicationStep in ProfileCreateFlight
- Notably, added this after the profile metadata was saved and made available as a job response in the working map
- Expanded docstrings, unit testing of AzureApplicationDeploymentService.getOrRegisterApplicationDeployment

okotsopoulos committed:
Debug failing tests: disable AzureAuthService caching

Test failures and flakiness seems to happen…
- During dataset creation…
- Specifically in CreateDatasetGetOrCreateStorageAccountStep when getting an existing SA (a new condition when billing profiles are shared, previously a new one was created with each test case)…
- Specifically in the call to azureBlobStorePdao.enableFileLogging

okotsopoulos committed:
Explicitly null out resource identifiers on set-up

Otherwise, test tear-down for a different test may unsuccessfully try to delete them -- observed for releaseSnapshotId specifically, which is not initialized in all tests.
